engine/access/rpc/backend/backend.go

```diff
@@ -226,12 +226,65 @@ func (b *Backend) GetNetworkParameters(_ context.Context) access.NetworkParamete
 
 // GetLatestProtocolStateSnapshot returns the latest finalized snapshot
 func (b *Backend) GetLatestProtocolStateSnapshot(_ context.Context) ([]byte, error) {
-	data, err := convert.SnapshotToBytes(b.state.Final())
+	snapshot := b.state.Final()
+
+	validSnapshot, err := b.getValidSnapshot(snapshot)
+	if err != nil {
+		return nil, err
+	}
+
+	return convert.SnapshotToBytes(validSnapshot)
+}
+
+func (b *Backend) getValidSnapshot(snapshot protocol.Snapshot) (protocol.Snapshot, error) {
```

Comment:
We should limit our iteration here to avoid both (1) expending a lot of time iterating in this endpoint and (2) sending snapshots with unusually large sealing segments over the wire. I think a reasonable limit (`SnapshotHistoryLimit`) would be ~50 blocks, since they typical sealing segment length is <10. 

We can define this limit more concretely as the maximum difference between the global latest finalized block height, and the block height of the lowest block in `snapshot.SealingSegment`, so it covers both cases.

Reply:
https://github.com/onflow/flow-go/pull/1940/commits/fe38afdec394fd3b9ac6caaf99023882dc95b68a